pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -304,11 +304,18 @@ public RebalanceResult rebalance(TableConfig tableConfig, Configuration rebalanc
     LOGGER.info("Rebalancing table: {} with minAvailableReplicas: {}, enableStrictReplicaGroup: {}, bestEfforts: {}",
         tableNameWithType, minAvailableReplicas, enableStrictReplicaGroup, bestEfforts);
     int expectedVersion = currentIdealState.getRecord().getVersion();
-    while (true) {
-      // Wait for ExternalView to converge before updating the next IdealState
-      IdealState idealState;
+
+    do {
+      Map<String, Map<String, String>> nextAssignment =
+          getNextAssignment(currentAssignment, targetAssignment, minAvailableReplicas, enableStrictReplicaGroup);
+      LOGGER.info("Got the next assignment for table: {} with number of segments to be moved to each instance: {}",
+          tableNameWithType,
+          SegmentAssignmentUtils.getNumSegmentsToBeMovedPerInstance(currentAssignment, nextAssignment));
+
+      Set<String> segmentsToMove = SegmentAssignmentUtils.getSegmentsToMove(currentAssignment, nextAssignment);
```

Comment:
`segmentsToMove` can be calculated with current and target assignment. We can keep the current logic of calculating next assignment.